src/main/java/uk/gov/ons/ctp/integration/rhsvc/service/impl/UniqueAccessCodeServiceImpl.java

```diff
@@ -101,4 +251,155 @@ private void sendRespondentAuthenticatedEvent(UniqueAccessCodeDTO data) throws C
+        caseTypeStr = request.getAddressType(); // again the equivalent
+      } else {
+        // 3rd choice. Use the case type from the UAC
+        caseTypeStr = uac.getCaseType();
+      }
+    }
+
+    CaseType caseType = CaseType.valueOf(caseTypeStr);
+
+    return caseType;
+  }
+
+  // Build a new case to store and send to RM via the NewAddressReported event.
+  private CollectionCase createCase(CaseType caseType, UAC uac, UACLinkRequestDTO request) {
+    CollectionCase newCase = new CollectionCase();
+
+    newCase.setId(UUID.randomUUID().toString());
+    newCase.setCollectionExerciseId(uac.getCollectionExerciseId());
+    newCase.setHandDelivery(false);
+    newCase.setSurvey("CENSUS");
+    newCase.setCaseType(caseType.name());
+
+    Address address = new Address();
+    address.setAddressLine1(request.getAddressLine1());
+    address.setAddressLine2(request.getAddressLine2());
+    address.setAddressLine3(request.getAddressLine3());
+    address.setTownName(request.getTownName());
+    address.setRegion(request.getRegion());
+    address.setPostcode(request.getPostcode());
+    address.setUprn(request.getUprn().asString());
+    address.setAddressType(caseType.name());
+    address.setEstabType(request.getEstabType());
+
+    // Set address level for case
+    if ((caseType == CaseType.CE || caseType == CaseType.SPG)
+        && uac.getFormType().equals(FormType.CE1.name())) {
+      address.setAddressLevel(AddressLevel.E.name());
+    } else {
+      address.setAddressLevel(AddressLevel.U.name());
+    }
+
+    newCase.setAddress(address);
+
+    log.with("caseId", newCase.getId())
+        .with("caseType", caseType)
+        .debug("Have populated CollectionCase object");
+
+    return newCase;
+  }
+
+  private void validateUACCase(UAC uac, CollectionCase collectionCase) throws CTPException {
+    // validate that the combination UAC.formType, UAC.caseType, Case.caseType are ALLOWED to be
+    // linked
+    // rather than disallowed. ie we will only link those combinations indicated as LINK:YES in the
+    // matrix included in
+    // https://collaborate2.ons.gov.uk/confluence/display/SDC/Auth.05+-+Unlinked+Authentication
+
+    FormType uacFormType = FormType.valueOf(uac.getFormType());
+    CaseType uacCaseType = CaseType.valueOf(uac.getCaseType());
+    CaseType caseCaseType = CaseType.valueOf(collectionCase.getCaseType());
+    Optional<LinkingCombination> linkCombo =
+        LinkingCombination.lookup(uacFormType, uacCaseType, caseCaseType);
+
+    if (linkCombo.isEmpty()) {
+      String failureDetails = uacFormType + ", " + uacCaseType + ", " + caseCaseType;
+      log.warn("Failed to link UAC to case. Incompatible combination: " + failureDetails);
+      throw new CTPException(
+          CTPException.Fault.BAD_REQUEST, "Case and UAC incompatible: " + failureDetails);
+    }
+  }
+
+  private UniqueAccessCodeDTO createUniqueAccessCodeDTO(UAC uac, CollectionCase collectionCase) {
```

Comment:
This could be used by the old code that you forgot - might need to make it null tolerant of course

Reply:
Have tweaked to be compatible with new usage